src/main/java/com/github/firmwehr/gentle/cli/CommandDispatcher.java

```diff
@@ -0,0 +1,43 @@
+package com.github.firmwehr.gentle.cli;
+
+import com.github.firmwehr.gentle.output.UserOutput;
+import com.github.firmwehr.gentle.util.Pair;
+
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+public class CommandDispatcher {
+	// TODO: Not a very pretty type. Perhaps this could be its own record.
+	private final List<Pair<Function<CommandArguments, Boolean>, Consumer<Path>>> commands;
+
+	public CommandDispatcher() {
+		this.commands = new ArrayList<>();
+	}
+
+	public CommandDispatcher command(Function<CommandArguments, Boolean> flag, Consumer<Path> command) {
+		this.commands.add(new Pair<>(flag, command));
+		return this;
+	}
+
+	public void dispatch(String[] args) {
+		CommandArguments arguments = new CommandArgumentsParser().parseOrExit(args);
+		List<Consumer<Path>> commands =
+			this.commands.stream().filter(cmd -> cmd.first().apply(arguments)).map(Pair::second).toList();
+
+		if (commands.size() == 0) {
+			UserOutput.userError("No operation specified.");
+			System.exit(1);
+		}
+
+		if (commands.size() > 1) {
+			// FIXME: Print conflicting flags.

```

Comment:
I don't think that is easily possible with your model as you do not have any information about the name of the flag. You'd need access to the `java.lang.reflect.Method` or sth like it.

Or you pass it in a parameter and store that in the record as well, which is probably a good compromise.

Reply:
Apparently there isn't really a neat way to access the `Method` instance when a method is used as a function. For now we'll just have to cope with a bit of redundancy.